### DISCUSSION TITLE
mirai, nanonext and NNG

### DISCUSSION THREAD

**Discussion by nviets** [ropensci/targets#783] [Ideas]
[This post](https://www.r-bloggers.com/2022/02/introducing-mirai-a-minimalist-async-evaluation-framework-for-r/) on Mirai on R Bloggers today looked like it might be interesting for targets' HPC frameworks. NNG is a [successor to zeromq](https://github.com/nanomsg/nng).

**Comment by wlandau** [comment]
Looks really cool, thanks @nviets. Maybe with enough development, `mirai` could have a role managing local async tasks in https://github.com/wlandau/crew and other task queues. FYI @krlmlr.

**Reply by nviets** [reply]
Crew looks awesome - can't wait to give it a try! I'm already a heavy users of clustermq for targets and other HPC workloads, and crew hits on a lot of important themes in core distributed workloads. Will crew allow for customization of heterogeneous workloads? Targets made it so easy to orchestrate sophisticated, asynchronous DAGs, and, sometimes, I want to assign specific worker configurations to tasks within the larger plans.

**Reply by wlandau** [reply]
The `main` branch at https://github.com/wlandau/crew is my first attempt, and it has a sketch of heterogeneity and worker tagging could look like. However, I quickly realized that that prototype would not perform well for cloud workers (e.g. AWS Batch). So now I am working on https://github.com/wlandau/crew/tree/11 to build up to asynchronous cloud worker management using a local multicore task queue. That async piece is a higher priority than heterogeneous workers, so I will engineer for it first and then loop back to heterogeneity later if I can.

By the way, `clustermq` is already a lot like `mirai`, and I tried to use it to build a local task queue much like :

* source: https://github.com/wlandau/crew/blob/11-clustermq/R/crew_queue_clustermq.R
* speed test: https://github.com/wlandau/crew/blob/11-clustermq/tests/performance/test-crew_queue_clustermq.R

It's not 100% working and it's slower than the [`callr`](https://www.tidyverse.org/blog/2019/09/callr-task-q/) version, but both problems could be on my end.

**Reply by nviets** [reply]
Hi @wlandau, any news or plans on your crew project? @shikokuchuo directed me to your exchange on persistent mirai sessions recently, and it reminded me of this thread. A big attraction of clustermq and future has always been that they plug into RM frameworks like slurm. Does crew aim for that? Does mirai?

**Reply by wlandau** [reply]
My goal for `crew` is to plug into cloud services (AWS Batch, Google Cloud Run, etc.) similarly to how `future.batchtools` and `clustermq` plug into traditional resource managers. Earlier this year, I put the project on hold because I realized I wanted to take the implementation in a completely different direction. This new direction will require a complete rewrite of the package.

The way I currently envision it, `crew` will be a hybrid-cloud task queue powered by @richfitz' [`rrq`](https://github.com/mrc-ide/rrq) package. [`rrq`](https://github.com/mrc-ide/rrq) already handles the message queue part, as well as complicated aspects of worker behavior and overall resilience. My plan for `crew` is to launch and manage R processes that run on the cloud and use [`rrq`](https://github.com/mrc-ide/rrq) to communicate and help with the life cycles of the handler and workers.

I am thinking the Redis server should run locally in a Docker container to make the logs easier to debug. This local container can also serve as a bastion to protect the local machine's IP etc. Then the cloud processes will have their own bastion job to isolate the workers and reduce the number of SSH connections (smaller attack surface). I also plan to encrypt (at least TLS), authenticate, and reduce the privileges of everything I can control. Security seems like at least half the battle.

I have been reading up on AWS and Docker in the months since I paused `crew`, and right now I am reading about computer networks and zero trust architecture. After covering those last two knowledge gaps, I think I will be ready for another attempt.

**Reply by nviets** [reply]
Thanks for the update and vision for what's next. Exciting stuff. AWS Parallel Cluster supports slurm, so it's straightforward to carry on-prem workloads directly into the cloud. I think GCP has something similar for slurm.

Will crew be specific to targets pipelines, or do you aim for something more comprehensive? Clustmermq is handy for distributing lower level tasks directly, e.g. loops and stuff.

rrq is new to me. Thanks for sharing. I'll read up. Sounds like you're at the leading edge of some foundational work in R.

**Reply by wlandau** [reply]
> AWS Parallel Cluster supports slurm, so it's straightforward to carry on-prem workloads directly into the cloud. I think GCP has something similar for slurm.

Right, I could just recommend using those services and `targets` via `tar_make_clustermq()` would just work. But it's old technology, and an entire cluster seems like a lot of centralized setup and teardown (could be a similar issue with Kubernetes). With `crew`, I want to aim for a hybrid cloud where all the remote HPC is ephemeral, hopefully quick to set up, and quick to exit when not needed.

> Will crew be specific to targets pipelines, or do you aim for something more comprehensive? Clustmermq is handy for distributing lower level tasks directly, e.g. loops and stuff.

If `crew` succeeds (I'll give it my best shot, but no guarantee) then it will be both. It will be a layer between `targets` and the cloud, and it will be usable on its own. For example, in Shiny apps this could be a heavy-duty alternative to `future`-powered `promises` when the number of long-running steps is sufficiently large. 

> Sounds like you're at the leading edge of some foundational work in R.

Thanks, we'll see. I have a lot left to learn, before, during, and after my next attempt.

**Reply by nviets** [reply]
Great to hear crew aims to handle both types of work.

This is probably my own bias speaking from working with clustermq on clusters so much, but how will crew/rrq match the dependency stack of R, libraries, and OS on the ephemeral workers? Wouldn't there would a high cost for shipping the dependencies to the worker process on a per task basis? In a cluster, everything can be installed and ready for when a slurm task lands on it.

**Reply by wlandau** [reply]
For package dependency installation, I think this will be handled through the Docker images that the workers will run.
